tests/sentry/event_manager/test_save_aggregate.py

```diff
@@ -0,0 +1,95 @@
+import contextlib
+import pytest
+import time
+
+from threading import Thread
+
+from sentry.eventstore.models import Event
+from sentry.event_manager import _save_aggregate, _save_aggregate2
+
+
+@pytest.mark.django_db(transaction=True)
+@pytest.mark.parametrize(
+    "save_aggregate_version",
+    [
+        # New group creation code, which is supposed to not have races
+        "new_save_aggregate",
+        # New group creation code with removed transaction isolation, which is then
+        # supposed to create multiple groups. This variant exists such that we can
+        # ensure the test would find race conditions in principle, and does not
+        # just always pass because of low parallelism. In a sense this variant
+        # tests the efficacy of this test, not actual business logic.
+        #
+        # If this variant fails, CONCURRENCY needs to be increased or e.g. thread
+        # barriers need to be used to ensure data races. This does not seem to be
+        # necessary so far.
+        "new_broken_save_aggregate",
+        # Old group creation code which is "supposed to" have races
+        "old_save_aggregate",
+    ],
+)
+def test_group_creation_race(monkeypatch, default_project, save_aggregate_version):
+    CONCURRENCY = 2
+
+    new_variant = save_aggregate_version in ("new_save_aggregate", "new_broken_save_aggregate")
+    is_race_free = save_aggregate_version == "new_save_aggregate"
+
+    if save_aggregate_version == "new_broken_save_aggregate":
+
+        class FakeTransactionModule:
+            @staticmethod
+            @contextlib.contextmanager
+            def atomic():
+                yield
+
+        # Disable transaction isolation just within event manager, but not in
+        # GroupHash.objects.create_or_update
+        monkeypatch.setattr("sentry.event_manager.transaction", FakeTransactionModule)
+
+        # select_for_update cannot be used outside of transactions
+        monkeypatch.setattr("django.db.models.QuerySet.select_for_update", lambda self: self)
```

Comment:
Is this right with `@pytest.mark.django_db(transaction=True)`. According to the docs it can be used with `django.test.TransactionTestCase`.
- https://docs.djangoproject.com/en/2.1/topics/testing/tools/#django.test.TransactionTestCase
- https://pytest-django.readthedocs.io/en/latest/helpers.html#pytest.mark.django_db

Reply:
TransactionTestCase means that you can open transactions in your app. Since we disable `transaction.atomic()` decorator using the preceding monkeypatch, we are never in a transaction. In the other test cases we do use select_for_update, but we can do that because we don't monkeypatch `with transaction.atomic()`

Reply:
Why do we need to patch the `transaction.atomic` since we are using the `django.test.TransactionTestCase`?